prow/cmd/webhook-server/helpers.go

```diff
@@ -246,37 +254,183 @@ func createValidatingWebhookConfig(ctx context.Context, caPem string, client ctr
+	operations := []admregistration.OperationType{"CREATE"}
+	scope := admregistration.ScopeType("*")
+	path := "/mutate"
+	sideEffects := admregistration.SideEffectClass("None")
+
+	mutatingWebhookConfig := &admregistration.MutatingWebhookConfiguration{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "MutatingWebhookConfiguration",
+			APIVersion: "admissionregistration.k8s.io/v1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: prowJobMutatingWebhookName,
+		},
+		Webhooks: []admregistration.MutatingWebhook{
+			{
+				Name: prowJobMutatingWebhookName,
+				ObjectSelector: &v1.LabelSelector{
+					MatchLabels: map[string]string{
+						"admission-webhook": "enabled",
+						"default-me":        "enabled",
+					},
+				},
+				Rules: []admregistration.RuleWithOperations{
+					{
+						Operations: operations,
+						Rule: admregistration.Rule{
+							APIGroups:   []string{"prow.k8s.io"},
+							APIVersions: []string{"v1"},
+							Resources:   []string{"prowjobs"},
+							Scope:       &scope,
+						},
+					},
+				},
+				ClientConfig: admregistration.WebhookClientConfig{
+					Service: &admregistration.ServiceReference{
+						Namespace: defaultNamespace,
+						Name:      prowjobAdmissionServiceName,
+						Path:      &path,
+					},
+					CABundle: []byte(caPem),
+				},
+				SideEffects:             &sideEffects,
+				AdmissionReviewVersions: []string{"v1"},
+			},
+		},
+	}
+
+	createOptions := &ctrlruntimeclient.CreateOptions{
+		FieldManager: "webhook-server",
+	}
+
+	err := client.Create(ctx, mutatingWebhookConfig, createOptions)
+	if err != nil && strings.Contains(err.Error(), configAlreadyExistsError) {
+		logrus.Info("MutatingWebhookConfiguration already exists, proceeding to patch")
+		if err := patchMutatingWebhookConfig(ctx, caPem, client); err != nil {
+			return fmt.Errorf("failed to patch mutating webhook config: %w", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("failed to create mutating webhook config: %w", err)
+	}
+
+	return nil
+}
+
+func patchMutatingWebhookConfig(ctx context.Context, caPem string, client ctrlruntimeclient.Client) error {
+	key := types.NamespacedName{
+		Namespace: defaultNamespace,
+		Name:      prowJobMutatingWebhookName,
+	}
+
+	patchOptions := &ctrlruntimeclient.PatchOptions{
+		FieldManager: "webhook-server",
+	}
+	var mutatingWebhookConfig admregistration.MutatingWebhookConfiguration
+	if err := client.Get(ctx, key, &mutatingWebhookConfig); err != nil {
+		return fmt.Errorf("failed to get mutating webhook config: %w", err)
+	}
+	oldMutatingWebhook := mutatingWebhookConfig.DeepCopy()
+	mutatingWebhookConfig.Webhooks[0].ClientConfig.CABundle = []byte(caPem)
+	if err := client.Patch(ctx, &mutatingWebhookConfig, ctrlruntimeclient.MergeFrom(oldMutatingWebhook), patchOptions); err != nil {
+		return fmt.Errorf("failed to patch mutating webhook config: %w", err)
+	}
+	return nil
+}
+
+func checkWebhooksExist(ctx context.Context, client ctrlruntimeclient.Client) (string, string, bool, error) {
```

Comment:
this function checks for existence of two configs, and returns only a single boolean, I'm not sure that I get the idea. Could you add a comment explaining why do this instead of have two separate function `checkValidatingWebhookExist` and `checkMutatingWebhookExist`, then let them return separately?

Reply:
ultimately even if only one was missing we would still have to reconcile both just to ensure there are no certificate errors arising. its much safer, so we can combined that check into one using a logical operation and then go from there.

Reply:
ack